### PR TITLE
[Feature](ipv6)Support IPV6

### PR DESCRIPTION
**Pull request by zhengshengjun**
# Proposed changes
Support IPV6 in Apache Doris, the main changes are:
1. enable binding to IPV6 address if network priority in config file contains an IPV6 CIDR string
2. BRPC and HTTP support binding to IPV6 address
3. BRPC and HTTP support visiting IPV6 Services

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [] No
    - [ ] I don't know
4. Has unit tests been added:
    - [] Yes
    - [ ] No
    - [ ] No Need
5. Has document been added or modified:
    - [ ] Yes
    - [] No
    - [ ] No Need
6. Does it need to update dependencies:
    - [] Yes
    - [ ] No
7. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/pom.xml [hunk 1]
```diff
@@ -181,7 +181,7 @@ under the License.
         <java-cup.version>0.11-a-czt02-cdh</java-cup.version>
         <javassist.version>3.18.2-GA</javassist.version>
         <javax.servlet-api.version>3.0.1</javax.servlet-api.version>
-        <je.version>18.3.13-doris-SNAPSHOT</je.version>
+        <je.version>18.3.14-doris-SNAPSHOT</je.version>
```
Comment (hf200012) [inline fe/pom.xml:184]:
```suggestion
        <je.version>18.3.13-doris-SNAPSHOT</je.version>
```

File: be/src/runtime/data_stream_sender.cpp [hunk 2]
```diff
@@ -164,8 +164,14 @@ Status DataStreamSender::Channel::send_batch(PRowBatch* batch, bool eos) {
                                                            RefCountClosure<PTransmitDataResult>>(
                 &_brpc_request, _closure);
         RETURN_IF_ERROR(st);
-        std::string brpc_url =
-                fmt::format("http://{}:{}", _brpc_dest_addr.hostname, _brpc_dest_addr.port);
+        std::string brpc_url;
+        //format an ipv6 address
```
Comment (morningman) [inline be/src/runtime/data_stream_sender.cpp:168]: I think we should extract a method for this.

File: be/src/service/brpc_service.cpp [hunk 3]
```diff
@@ -49,7 +49,18 @@ Status BRpcService::start(int port, int num_threads) {
         options.num_threads = num_threads;
     }
 
-    if (_server->Start(port, &options) != 0) {
+    butil::EndPoint point;
+    const char* host;
```
Comment (morningman) [inline be/src/service/brpc_service.cpp:53]: extract a method

File: be/src/util/brpc_client_cache.h [hunk 4]
```diff
@@ -60,7 +60,17 @@ class BrpcClientCache {
...
     std::shared_ptr<T> get_client(const std::string& host, int port) {
         std::string host_port = fmt::format("{}:{}", host, port);
-        return get_client(host_port);
+        std::shared_ptr<T> stub_ptr;
+        auto get_value = [&stub_ptr](typename StubMap<T>::mapped_type& v) { stub_ptr = v; };
+        if (LIKELY(_stub_map.if_contains(host_port, get_value))) {
+            return stub_ptr;
+        }
+
+        // new one stub and insert into map
```
Comment (morningman) [inline be/src/util/brpc_client_cache.h:69]: duplicate code

File: fe/fe-core/src/main/java/org/apache/doris/qe/HelpModule.java [hunk 5]
```diff
@@ -266,7 +266,8 @@ public void setUpModule() throws IOException, UserException {
         URL helpResource = instance.getClass().getClassLoader()
                 .getResource(HELP_ZIP_FILE_NAME);
         if (helpResource == null) {
-            throw new IOException("Can not find help zip file: " + HELP_ZIP_FILE_NAME);
+            return;
+            // throw new IOException("Can not find help zip file: " + HELP_ZIP_FILE_NAME);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/qe/HelpModule.java:270]: Why modify this?

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.71 seconds
 load time: 475 seconds
 storage size: 17122864446 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230108165541_clickbench_pr_75725.html

File: fe/fe-core/src/main/java/org/apache/doris/httpv2/controller/BaseController.java [hunk 6]
```diff
@@ -296,6 +296,6 @@ protected long checkLongParam(String strParam) {
     }
 
     protected String getCurrentFrontendURL() {
-        return "http://" + FrontendOptions.getLocalHostAddress() + ":" + Config.http_port;
+        return "http://" + FrontendOptions.getHostname() + ":" + Config.http_port;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/httpv2/controller/BaseController.java:299]: What is diff between `getLocalHostAddress` and `getHostname`?

File: be/src/util/network_util.cpp [hunk 7]
```diff
@@ -39,22 +39,38 @@
...
-    return addr.sin_family == AF_INET;
+bool InetAddress::is_loopback() {
+    if (addr.sa_family == AF_INET) {
+        in_addr_t s_addr = ((struct sockaddr_in*)&addr)->sin_addr.s_addr;
+        return (ntohl(s_addr) & 0xFF000000) == 0x7F000000;
+    } else if (addr.sa_family == AF_INET6) {
+        struct in6_addr in6_addr = ((struct sockaddr_in6*)&addr)->sin6_addr;
+        return IN6_IS_ADDR_LOOPBACK(&in6_addr);
+    } else {
+        LOG(WARNING) << "unknow address";
```
Comment (morningman) [inline be/src/util/network_util.cpp:53]:
```suggestion
        LOG(WARNING) << "unknow address: " << addr.sa_family;
```

File: be/src/util/network_util.cpp [hunk 8]
```diff
@@ -39,22 +39,38 @@
...
+        char addr_buf[INET_ADDRSTRLEN];
+        inet_ntop(AF_INET, &(((struct sockaddr_in*)&addr)->sin_addr), addr_buf, INET_ADDRSTRLEN);
+        return std::string(addr_buf);
+    } else if (addr.sa_family == AF_INET6) {
+        char addr_buf[INET6_ADDRSTRLEN];
+        inet_ntop(AF_INET6, &(((struct sockaddr_in6*)&addr)->sin6_addr), addr_buf,
+                  INET6_ADDRSTRLEN);
+        return std::string(addr_buf);
+    } else {
+        return std::string {"unknown address"};
```
Comment (morningman) [inline be/src/util/network_util.cpp:69]:
```suggestion
        return std::string {"unknown address: " + addr.sa_family};
```

**Review by morningman — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**